src/Component.ts

```diff
@@ -0,0 +1,310 @@
+type EventHash<T extends EventMap> = Partial<{ [K in EventKey<T>]: EventCallback<T, K> }>;
+
+type EventCallback<T extends EventMap, K extends EventKey<T>>
+  = T[K] extends NoArguments
+    ? (event: DefaultParam<T>) => any
+    : T[K] extends (evt: infer U, ...restParam: infer V) => any
+      ? (event: U & DefaultParam<T>, ...restParam: V) => any
+      : (event: T[K] & DefaultParam<T>) => any;
+type EventParam<T extends EventMap, K extends EventKey<T>>
+  = T[K] extends NoArguments
+    ? (event: DefaultParam<T>) => any
+    : T[K] extends EventWithRestParam
+      ? T[K]
+      : (event: T[K]) => any;
+
+/**
+ * A class used to manage events in a component
+ * @ko 컴포넌트의 이벤트을 관리할 수 있게 하는 클래스
+ * @alias eg.Component
+ */
+class Component<T extends EventMap> {
+  /**
+   * Version info string
+   * @ko 버전정보 문자열
+   * @name VERSION
+   * @static
+   * @type {String}
+   * @example
+   * eg.Component.VERSION;  // ex) 2.0.0
+   * @memberof eg.Component
+   */
+  public static VERSION: string = "#__VERSION__#";
+
+  /**
+   * @deprecated
+   */
+  public options: {[key: string]: any} = {};
+  private _eventHandler: {[keys: string]: EventCallback<T, EventKey<T>>[]};
+
+  /**
+   * @support {"ie": "7+", "ch" : "latest", "ff" : "latest",  "sf" : "latest", "edge" : "latest", "ios" : "7+", "an" : "2.1+ (except 3.x)"}
+   */
+  constructor() {
+    this._eventHandler = {};
+  }
+
+  /**
+   * Triggers a custom event.
+   * @ko 커스텀 이벤트를 발생시킨다
+   * @param {String} eventName The name of the custom event to be triggered <ko>발생할 커스텀 이벤트의 이름</ko>
+   * @param {Object} customEvent Event data to be sent when triggering a custom event <ko>커스텀 이벤트가 발생할 때 전달할 데이터</ko>
+   * @return {Boolean} Indicates whether the event has occurred. If the stop() method is called by a custom event handler, it will return false and prevent the event from occurring. <a href="https://github.com/naver/egjs-component/wiki/How-to-make-Component-event-design%3F">Ref</a> <ko>이벤트 발생 여부. 커스텀 이벤트 핸들러에서 stop() 메서드를 호출하면 'false'를 반환하고 이벤트 발생을 중단한다. <a href="https://github.com/naver/egjs-component/wiki/How-to-make-Component-event-design%3F">참고</a></ko>
+   * @example
+   * class Some extends eg.Component {
+   *   some(){
+   *     if(this.trigger("beforeHi")){ // When event call to stop return false.
+   *       this.trigger("hi");// fire hi event.
+   *     }
+   *   }
+   * }
+   *
+   * const some = new Some();
+   * some.on("beforeHi", (e) => {
+   *   if(condition){
+   *     e.stop(); // When event call to stop, `hi` event not call.
+   *   }
+   * });
+   * some.on("hi", (e) => {
+   *   // `currentTarget` is component instance.
+   *   console.log(some === e.currentTarget); // true
+   * });
+   * // If you want to more know event design. You can see article.
+   * // https://github.com/naver/egjs-component/wiki/How-to-make-Component-event-design%3F
+   */
+  public trigger<K extends EventKey<T>>(eventName: K, customEvent: Partial<T[K]> = {}, ...restParam): boolean {
```

Comment:
@WoodNeck 
It seems that the eventParam value to be triggered should not be partial.

Reply:
Well I thought it should be partial while I'm changing that function into Typescript, as the original code always allows `customEvent` to be just an empty object(`{}`)
But you're right, I think that is wrong way to use Component and should be rewritten.
Maybe no major version update is needed as it'll show compile error.

Reply:
To avoid errors, you have to do not set defaultValue.

```ts
trigger<>(e: K, customEvent?: FirstParam<T, K>)
```

The following code is also checking for customEvent.

Reply:
Well, I said error is just okay as users should change their original code.

Reply:
@WoodNeck 
But trigger must also be 1 argument.

```js
a.trigger("event1");
```


how about overrides
```ts
type Values<T> = T[keyof T];
type EventNoParamKey<T extends EventMap> = Values<{
  [key in keyof T]: {} extends T[key] ? key : never;
}> & string;
public trigger(eventName: EventNoParamKey<T>): boolean;
public trigger<K extends EventKey<T>>(eventName: K, customEvent: FirstParam<T, K> | never | undefined, ...restParam: RestParam<T, K>): boolean;
/**
*/
public trigger(eventName: string, customEvent?: any, ...restParam: any[]): boolean {
//....
}
```

Reply:
Okay, I think that should be revised.